opm/common/OpmLog/MessageLimiter.hpp

```diff
@@ -46,62 +54,108 @@ namespace Opm
         /// Negative limits (including NoLimit) are interpreted as
         /// NoLimit, but the default constructor is the preferred way
         /// to obtain that behaviour.
-        explicit MessageLimiter(const int message_limit)
-            : message_limit_(message_limit < 0 ? NoLimit : message_limit)
+        explicit MessageLimiter(const int tag_limit)
+            : tag_limit_(tag_limit < 0 ? NoLimit : tag_limit),
+              category_limits_({{Log::MessageType::Note, NoLimit},
+                                {Log::MessageType::Info, NoLimit},
+                                {Log::MessageType::Warning, NoLimit},
+                                {Log::MessageType::Error, NoLimit},
+                                {Log::MessageType::Problem, NoLimit},
+                                {Log::MessageType::Bug, NoLimit}})
+        {
+        }
+
+        MessageLimiter(const int tag_limit, const std::map<int64_t, int> category_limits)
+            : tag_limit_(tag_limit < 0 ? NoLimit : tag_limit),
+              category_limits_(category_limits)
         {
         }
 
         /// The message limit (same for all tags).
         int messageLimit() const
         {
-            return message_limit_;
+            return tag_limit_;
         }
 
-        /// Used for handleMessageTag() return type (see that
+        /// Used for handleMessageLimits() return type (see that
         /// function).
         enum class Response
         {
-            PrintMessage, JustOverLimit, OverLimit
+            PrintMessage, JustOverTagLimit, JustOverCategoryLimit, OverTagLimit, OverCategoryLimit
         };
 
         /// If a tag is empty, there is no message limit or for that
-        /// tag (count <= limit), respond PrintMessage.
+        /// (count <= limit), respond PrintMessage.
         /// If (count == limit + 1), respond JustOverLimit.
         /// If (count > limit + 1), respond OverLimit.
-        Response handleMessageTag(const std::string& tag)
+        Response handleMessageLimits(const std::string& tag, const int64_t messageMask)
         {
-            if (tag.empty() || message_limit_ == NoLimit) {
-                return Response::PrintMessage;
+            Response res = Response::PrintMessage;
+            // Deal with tag limits.
+            if (tag.empty() || tag_limit_ == NoLimit) {
+                category_counts_[messageMask]++;
```

Comment:
Prefer the simpler and more intuitive pre-increment to post-increment (unless necessary for algorithm logic).

(Also below.)